TestMaxInFlight is flaky

I've already seen it happening few times:

    E1001 13:18:21.582768    6716 errors.go:62] apiserver received an error that is not an unversioned.Status: test Error
    E1001 13:18:21.758605    6716 apiserver.go:453] Failed to parse "not a timeout": time: invalid duration not a timeout
    E1001 13:18:21.782217    6716 apiserver.go:421] apiserver was unable to write a JSON response: type apiserver.UnregisteredAPIObject is not registered for "version" and it will be impossible to Decode it, therefore Encode will refuse to encode it.
    E1001 13:18:21.782232    6716 errors.go:62] apiserver received an error that is not an unversioned.Status: type apiserver.UnregisteredAPIObject is not registered for "version" and it will be impossible to Decode it, therefore Encode will refuse to encode it.
    --- FAIL: TestMaxInFlight (0.00s)
    	handlers_test.go:49: unexpected response: 429
    FAIL
    FAIL	k8s.io/kubernetes/pkg/apiserver	0.466s

--- a/pkg/apiserver/handlers_test.go
+++ b/pkg/apiserver/handlers_test.go
@@ -62,6 +62,9 @@ func TestMaxInFlight(t *testing.T) {
 	const Iterations = 3
 	block := sync.WaitGroup{}
 	block.Add(1)
+	oneFinished := sync.WaitGroup{}
+	oneFinished.Add(1)
+	var once sync.Once
 	sem := make(chan bool, Iterations)
 
 	re := regexp.MustCompile("[.*\\/watch][^\\/proxy.*]")
@@ -86,6 +89,7 @@ func TestMaxInFlight(t *testing.T) {
 		// These should hang waiting on block...
 		go func() {
 			expectHTTP(server.URL+"/foo/bar/watch", http.StatusOK, t)
+			once.Do(oneFinished.Done)
 		}()
 	}
 
@@ -93,6 +97,7 @@ func TestMaxInFlight(t *testing.T) {
 		// These should hang waiting on block...
 		go func() {
 			expectHTTP(server.URL+"/proxy/foo/bar", http.StatusOK, t)
+			once.Do(oneFinished.Done)
 		}()
 	}
 	expectHTTP(server.URL+"/dontwait", http.StatusOK, t)
@@ -101,6 +106,7 @@ func TestMaxInFlight(t *testing.T) {
 		// These should hang waiting on block...
 		go func() {
 			expectHTTP(server.URL, http.StatusOK, t)
+			once.Do(oneFinished.Done)
 		}()
 	}
 	calls.Wait()
@@ -117,6 +123,8 @@ func TestMaxInFlight(t *testing.T) {
 	block.Done()
 
 	// Show that we recover from being blocked up.
+	// However, we should until at least one of the requests really finishes.
+	oneFinished.Wait()
 	expectHTTP(server.URL, http.StatusOK, t)
 }
 

